lesson 5/OpenCVtest/tensor.h

```diff
@@ -0,0 +1,108 @@
+#pragma once
+#include <iostream>
+#include <vector>
+#include <initializer_list>
+#include <algorithm>
+#include <array>
+#include <opencv2/core/mat.hpp>
+#include <opencv2/core/core.hpp>
+
+enum Orientation { HORIZONTAL, VERTICAL };
+
+class Tensor {
+public:
+	virtual ~Tensor() noexcept {}
+	virtual std::array<std::size_t, 2> getShape() const = 0;
+	virtual std::vector<double> flatten() const = 0;
+	virtual void transpose() = 0;
+	virtual void printElements() const = 0;
+};
+
+class Tensor1D : public Tensor {
+	double* data;
+	std::size_t size;
+	std::array<std::size_t, 2> shape;
+public:
+	Tensor1D() : data{ nullptr }, size{ 0 }, shape{ 0 }{}
+	Tensor1D(std::initializer_list<double>, Orientation m_n_ = HORIZONTAL);
+	virtual ~Tensor1D() noexcept;
+
+	Tensor1D(const Tensor1D&);
+	Tensor1D(Tensor1D&&);
+	void printElements() const override;
+	Tensor1D& operator=(const Tensor1D&);
+	Tensor1D& operator=(Tensor1D&&);
+
+	std::array<std::size_t, 2> getShape() const override;
+	std::vector<double> flatten() const override;
+	void transpose() override;
+	friend void swap(Tensor1D&, Tensor1D&);
+};
+
+class Tensor2D : public Tensor {
+	double** data;
+	std::array<std::size_t, 2> shape;
+
+	void allocate2D(const std::size_t, const std::size_t);
+	void allocate2D(const std::size_t, std::size_t, const double);
+
```

Comment:
add here another overloaded function `allocate2D`
`template <typename destIterator>
	void allocate2D(const std::size_t rowSize, const std::size_t columnSize, destIterator it);`
	
Which will both allocate and initialize elements with the given iterator